Fix AbstractDateField parsing and errors handling, support locale

elmot opened the pull request:
Fixes #9518
Fixes #8991
Fixes #8687

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/9740)
<!-- Reviewable:end -->

Commit:
Fix AbstractDateField parsing and errors handling, support locale

Fixes #9518
Fixes #8991
Fixes #8687

Commit:
Forgotten change

hesara commented:
Review status: 0 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 215 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq1ywZZ5cpBS_IAu4qW:-Kq1ywZZ5cpBS_IAu4qX:b-nkw1v5) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L215)):*
> ```Java
>             T newDate;
> 
>             if("".equals(newDateString)) {
> ```

this assumes newDateString is not null, but code about 12 lines down checks for null - can this be null?

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 230 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq20NHfuqCd9hCzJLm7:-Kq20NHfuqCd9hCzJLm8:b-bah7jq) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L230)):*
> ```Java
> 
>                 if (newDateString != null && !newDateString.isEmpty()) {
>                     String invalidDateString = (String) variables.get("lastInvalidDateString");
> ```

I'm not 100% sure if this is set in the correct situations (e.g. would the client and the server parse the date differently in some situations?) - to check in VAbstractTextualDate

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 235 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq20zZl_6F27DAlO1IH:-Kq20zZl_6F27DAlO1II:bdlrbs6) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L235)):*
> ```Java
>                         uiHasValidDateString = false;
>                         currentParseErrorMessage = parsedDate.getMessage().orElse("Parsing error");
>                         setComponentError(new UserError(getParseErrorMessage()));
> ```

What will getValue() return in this case? Should the release notes mention it if the behavior has changed? Will there be events?

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 242 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq1zTtgr2hDkKaL9F2f:-Kq1zTtgr2hDkKaL9F2g:b-fc7wnt) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L242)):*
> ```Java
>                     }
>                 } else {
>                     setValue(newDate,true);
> ```

(here and above, missing space after comma)

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 549 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq21PcDCUn_aERgEJ_i:-Kq21PcDCUn_aERgEJ_j:bh7sg9w) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L549)):*
> ```Java
>                                     + " as date");
>                 }
>                 dateString = formatDate(value);
> ```

Are the designs always using some specific locale? Or component locale?

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 560 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq21V_uJhjeb_K7kwKZ:-Kq21V_uJhjeb_K7kwK_:b-w6noed) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L560)):*
> ```Java
> 
>     /**
>      * Formats date according to the components locale.
> ```

see the comment above

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 611 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq21ky_4X8IJNYA8gQT:-Kq21ky_4X8IJNYA8gQU:bruhcb2) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L611)):*
> ```Java
>             dateString = formatDate(value);
>         } else {
>             dateString = formatDate(getEmptyValue());
> ```

Does this change the behavior?

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 620 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq21xXrLoZ4Ssyz008N:-Kq21xXrLoZ4Ssyz008O:bwzzvj9) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L620)):*
> ```Java
>                     new ValueContext(this, this));
>             if (result.isError()) {
>                 currentParseErrorMessage = getDateOutOfRangeMessage();
> ```

I guess this almost corresponds to the old behavior, but not quite as uiHasValidDateString is not modified - does that have some implications?

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 708 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq22MBItRM4rC0pMOY0:-Kq22MBItRM4rC0pMOY1:b-ht6fo) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L708)):*
> ```Java
>     @Override
>     public Validator<T> getDefaultValidator() {
>         return (Validator<T>) (value, context) -> {
> ```

I wonder if this change to lambda might (or might not) break serialization - see e.g. https://github.com/vaadin/framework/pull/9721

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: 0 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 215 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq1ywZZ5cpBS_IAu4qW:-Kq2JzruiH1aXMEBRowB:btd6cmt) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L215)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

this assumes newDateString is not null, but code about 12 lines down checks for null - can this be null?
</blockquote></details>

Ok, let's keep it safe

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 230 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq20NHfuqCd9hCzJLm7:-Kq2NA29Q2Rl68kfGfUT:b-ej08fi) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L230)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

I'm not 100% sure if this is set in the correct situations (e.g. would the client and the server parse the date differently in some situations?) - to check in VAbstractTextualDate
</blockquote></details>

The code is changed according to the documentation (and that doc is not 100% correct, i will fix that a bit later)

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 235 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq20zZl_6F27DAlO1IH:-Kq2KHF3TEkGxpfZ_IS5:bfckkrt) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L235)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

What will getValue() return in this case? Should the release notes mention it if the behavior has changed? Will there be events?
</blockquote></details>

Old value will be kept if user typed in some wrong text.

Old behavior looked completely broken to me, so it looks more like fix rather than behavior change.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 242 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq1zTtgr2hDkKaL9F2f:-Kq2KdX4ebRSwrCzhPAn:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L242)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(here and above, missing space after comma)
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 549 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq21PcDCUn_aERgEJ_i:-Kq2KzCW2WJ86PGDgYSh:bhvmwnc) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L549)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Are the designs always using some specific locale? Or component locale?
</blockquote></details>

Let's fix that later.
Reverted, and I added an issue #9744. 

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 560 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq21V_uJhjeb_K7kwKZ:-Kq2MUjHRFMO-zvXQ_Ua:b-9kngae) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L560)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

see the comment above
</blockquote></details>

Not used during design reads anymore

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 611 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq21ky_4X8IJNYA8gQT:-Kq23nDh4JOeRBSYxQEu:b-z62i7v) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L611)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Does this change the behavior?
</blockquote></details>

Not really.  Empty date is converted to empty  string for the moment. And with this change the component subclasses might respect empty values some other way.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 620 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq21xXrLoZ4Ssyz008N:-Kq2MzM_OVkY93kC6Ybe:b-gremws) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L620)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

I guess this almost corresponds to the old behavior, but not quite as uiHasValidDateString is not modified - does that have some implications?
</blockquote></details>

That did not work properly, and I could not find any reasonable implication.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 708 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq22MBItRM4rC0pMOY0:-Kq252plxg-bAYijdF5j:bw68pgw) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L708)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

I wonder if this change to lambda might (or might not) break serialization - see e.g. https://github.com/vaadin/framework/pull/9721
</blockquote></details>

Let's keep that safe. Reverted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

Commit:
Proper error clean, server-side parsing, fixes after review.

Commit:
Fix invalid text handling

Commit:
Empty string handling fix

Commit:
Empty string handling fix again

Commit:
Fix tests

Commit:
Fix test

Commit:
One more test

Commit:
Fix documentation

Artur- commented:
Review status: 0 of 11 files reviewed at latest revision, 9 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 235 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq20zZl_6F27DAlO1IH:-KqMV_VJpggqcEqc2guL:b-fwzvk) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L235)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Old value will be kept if user typed in some wrong text.

Old behavior looked completely broken to me, so it looks more like fix rather than behavior change.
</blockquote></details>

Why should the behavior change? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

Artur- commented:
Review status: 0 of 11 files reviewed at latest revision, 9 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 235 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq20zZl_6F27DAlO1IH:-KqMVg7c8rBrD6Mo6IM7:b-dvjw1v) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L235)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Why should the behavior change? 
</blockquote></details>

As far as I remember, earlier the value was set to null so you would not end up with
1. User enters "2015-01-01"
2. User enters "foo"
3. User saves form -> "2015-01-01" is stored even though user can't possibly expect it to be

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: 0 of 11 files reviewed at latest revision, 7 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 235 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq20zZl_6F27DAlO1IH:-KqNQPpg92quyM7nPK0Y:b-7p3i79) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L235)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

As far as I remember, earlier the value was set to null so you would not end up with
1. User enters "2015-01-01"
2. User enters "foo"
3. User saves form -> "2015-01-01" is stored even though user can't possibly expect it to be
</blockquote></details>

As far as I understand the form can\t be submitted until internal error is cleared by entering correct date by typing or by choosing from the calendar. Keeping old value in the field shows previous date in the calendar popup even if user typed something incorrect, which is UX improvement, imho. @anasmi, what do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: 0 of 11 files reviewed at latest revision, 7 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 215 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq1ywZZ5cpBS_IAu4qW:-KqNRFayrkYbBgAORNMM:bab1dfv) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L215)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Ok, let's keep it safe
</blockquote></details>

I was wrong. It should be tested against empty string. Here is a difference if user has cleared textbox or choosen date from popup.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

Commit:
Formatting fix

hesara commented:
Reviewed 1 of 5 files at r1, 5 of 9 files at r2.
Review status: 6 of 11 files reviewed at latest revision, 16 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 548 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRSuwcWLVLZrszcvXJ:-KqRSuwdA7mJjwtYbcGG:b-yy9z27) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L548)):*
> ```Java
>      * @return textual representation of the date or empty string for {@code null}
>      */
>     protected abstract String formatDate(T value);
> ```

The new abstract method (which is also missing a since tag) is a breaking change for anyone extending ADF - perhaps a default implementation would help.

---

*[server/src/main/java/com/vaadin/ui/AbstractLocalDateField.java, line 143 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRTDNAQsPCM4Oi_qlF:-KqRTDNAQsPCM4Oi_qlG:b-4nn6hk) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/server/src/main/java/com/vaadin/ui/AbstractLocalDateField.java#L143)):*
> ```Java
> 
>     @Override
>     protected String formatDate(LocalDate value) {
> ```

Move as default implementation to superclass?

---

*[server/src/main/java/com/vaadin/ui/AbstractLocalDateTimeField.java, line 171 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRTIDlhA6yyoEVGYqo:-KqRTIDlhA6yyoEVGYqp:b-4nn6hk) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/server/src/main/java/com/vaadin/ui/AbstractLocalDateTimeField.java#L171)):*
> ```Java
> 
>     @Override
>     protected String formatDate(LocalDateTime value) {
> ```

Move as default implementation to superclass?

---

*[uitest/src/main/java/com/vaadin/tests/components/datefield/DateTextHandling.java, line 48 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRUCflhiEOxI7f1BO1:-KqRUCflhiEOxI7f1BO2:b-tnb9wl) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/main/java/com/vaadin/tests/components/datefield/DateTextHandling.java#L48)):*
> ```Java
> 
>         Binder<Void> binder = new Binder<>();
>         binder.forField(dateField).withStatusLabel(errorLabel).bind(o -> dateField.getEmptyValue(), null);
> ```

always returning empty value?

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateFieldIsValidTest.java, line 49 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRVrNWGzCbEfeKQFqh:-KqRVrNWGzCbEfeKQFqi:brid5c9) ([raw file](https://github.com/vaadin/framework/blob/94f2e1f4227c59b21011747ab04ed553cfa235a8/uitest/src/test/java/com/vaadin/tests/components/datefield/DateFieldIsValidTest.java#L49)):*
> ```Java
> 
>         dateTextbox.sendKeys("lala", Keys.TAB);
>         assertLogText("3. valueChange: value: null, is valid: false");
> ```

Was it an intended change that there is no VCE for invalid value? (My impression is that it was the intended behavior in the past anyway.)

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java, line 23 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRUuA7hBVDsU5jBVkN:-KqRUuA8eyBUcfj6wkAC:bwwue6z) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java#L23)):*
> ```Java
>         LabelElement validateResult = $(LabelElement.class).first();
>         WebElement dateTextbox = dateFieldElement
>                 .findElement(com.vaadin.testbench.By.className("v-textfield"));
> ```

isn't this the same as dateFieldElement.getInputElement() ?

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java, line 27 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRV-_Ay4us1cH5orxv:-KqRV-_Ay4us1cH5orxw:b-ovv6o9) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java#L27)):*
> ```Java
>         dateTextbox.sendKeys("Y2K",Keys.TAB);
>         validate.click();
>         assertNotification("Y2K Sould be converted to 1-JAN-2000", "01-Jan-2000");
> ```

Sould -> should, and could use the same cases in 1-Jan-2000 in the message

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateTimeFieldIsValidTest.java, line 49 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRVdGjKpI0_GO2Lrq_:-KqRVdGjKpI0_GO2Lrqa:brid5c9) ([raw file](https://github.com/vaadin/framework/blob/94f2e1f4227c59b21011747ab04ed553cfa235a8/uitest/src/test/java/com/vaadin/tests/components/datefield/DateTimeFieldIsValidTest.java#L49)):*
> ```Java
> 
>         dateTextbox.sendKeys("lala", Keys.TAB);
>         assertLogText("3. valueChange: value: null, is valid: false");
> ```

Was it an intended change that there is no VCE for invalid value? (My impression is that it was the intended behavior in the past anyway.)

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DisabledParentLayoutTest.java, line 68 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRVJCqQp7-hMxB05YR:-KqRVJCqQp7-hMxB05YS:bm5o90n) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/components/datefield/DisabledParentLayoutTest.java#L68)):*
> ```Java
> 
>         dataFieldButton.click();
>         dataFieldButton.click();//Requires two clicks because of error message. TODO fix
> ```

will the TODO be addressed?

---

*[uitest/src/test/java/com/vaadin/tests/elements/abstracttextfield/AbstractTextElementSetValueTest.java, line 67 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRVN2qsFqkVnJps-O2:-KqRVN2qsFqkVnJps-O3:b-sax8j5) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/elements/abstracttextfield/AbstractTextElementSetValueTest.java#L67)):*
> ```Java
>         // we can type any string in date field element
>         elem.setValue(TYPED_STRING);
>         // invalid values are cleared stays
> ```

not sure if I understand the comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

Commit:
Null handling, review fixes

elmot commented:
Review status: 2 of 9 files reviewed at latest revision, 14 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 548 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRSuwcWLVLZrszcvXJ:-KqS0VtVoSZIMLX-jgoe:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L548)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

The new abstract method (which is also missing a since tag) is a breaking change for anyone extending ADF - perhaps a default implementation would help.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/AbstractLocalDateField.java, line 143 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRTDNAQsPCM4Oi_qlF:-KqS0XM3AVCvdtWlZtIj:b-ctk2xm) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/server/src/main/java/com/vaadin/ui/AbstractLocalDateField.java#L143)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Move as default implementation to superclass?
</blockquote></details>

Impossible due generics in superclass

---

*[server/src/main/java/com/vaadin/ui/AbstractLocalDateTimeField.java, line 171 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRTIDlhA6yyoEVGYqo:-KqS0ctEabmUKAM-EAYP:b-ctk2xm) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/server/src/main/java/com/vaadin/ui/AbstractLocalDateTimeField.java#L171)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Move as default implementation to superclass?
</blockquote></details>

Impossible due generics in superclass

---

*[uitest/src/main/java/com/vaadin/tests/components/datefield/DateTextHandling.java, line 48 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRUCflhiEOxI7f1BO1:-KqS1vr8Hq5S7Svas7Ka:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/main/java/com/vaadin/tests/components/datefield/DateTextHandling.java#L48)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

always returning empty value?
</blockquote></details>

Done.

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java, line 23 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRUuA7hBVDsU5jBVkN:-KqS0xCZSaJ4BJjuRH0Z:byfoccg) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java#L23)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

isn't this the same as dateFieldElement.getInputElement() ?
</blockquote></details>

I don't see such a method

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java, line 27 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRV-_Ay4us1cH5orxv:-KqS1QOkkuZ4TOOxyJ4C:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java#L27)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Sould -> should, and could use the same cases in 1-Jan-2000 in the message
</blockquote></details>

Done.

---

*[uitest/src/test/java/com/vaadin/tests/elements/abstracttextfield/AbstractTextElementSetValueTest.java, line 67 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRVN2qsFqkVnJps-O2:-KqS1g0Vzh9mwQMdpAxj:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/elements/abstracttextfield/AbstractTextElementSetValueTest.java#L67)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

not sure if I understand the comment
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 2 of 5 files at r1, 2 of 6 files at r3.
Review status: 6 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java, line 23 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRUuA7hBVDsU5jBVkN:-KqSXZtHvpVvT1tDd25x:b4gzvhk) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java#L23)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

I don't see such a method
</blockquote></details>

Method added in master - rebasing this change would allow using it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

Commit:
Test revert

hesara commented:
Reviewed 2 of 6 files at r3.
Review status: 7 of 9 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 219 at r3](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqSjGS1Cyswb8vXFJ2T:-KqSjGS2WpfLsPqDDCtW:bwyglu9) ([raw file](https://github.com/vaadin/framework/blob/1d4dcdb8abd3c034de54c3d757cc2a1340596df2/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L219)):*
> ```Java
>             if ("".equals(newDateString)) {
>                 newDate = null;
>                 hasChanges = !uiHasValidDateString;
> ```

I'm not sure I understand this line - what does it do and why? - apparently handled by hasChanges modification below. Perhaps even this isn't needed with the hasChanges line below, maybe with an extra condition.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 220 at r3](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqSlvyCvWIUJAHJV7xd:-KqSlvyCvWIUJAHJV7xe:b-14ww8n) ([raw file](https://github.com/vaadin/framework/blob/1d4dcdb8abd3c034de54c3d757cc2a1340596df2/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L220)):*
> ```Java
>                 newDate = null;
>                 hasChanges = !uiHasValidDateString;
>                 uiHasValidDateString = true;
> ```

this line and the next might be unnecessary - add a TODO comment to check it?

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 610 at r3](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqSndUMJeJq7RABK4Eq:-KqSndUuhi0l953ebjU3:b-31kwse) ([raw file](https://github.com/vaadin/framework/blob/1d4dcdb8abd3c034de54c3d757cc2a1340596df2/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L610)):*
> ```Java
>         }
>         if (currentParseErrorMessage == null) {
>             setComponentError(null);
> ```

probably should also set uiHasValidDateString when the server calls setValue() and everything is ok

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

Commit:
After review changes

elmot commented:
Review status: 4 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 219 at r3](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqSjGS1Cyswb8vXFJ2T:-KqSriGQd7BXAQilS9QR:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/1d4dcdb8abd3c034de54c3d757cc2a1340596df2/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L219)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

I'm not sure I understand this line - what does it do and why? - apparently handled by hasChanges modification below. Perhaps even this isn't needed with the hasChanges line below, maybe with an extra condition.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 220 at r3](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqSlvyCvWIUJAHJV7xd:-KqSrkoIweDCYJohJTWM:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/1d4dcdb8abd3c034de54c3d757cc2a1340596df2/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L220)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

this line and the next might be unnecessary - add a TODO comment to check it?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 610 at r3](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqSndUMJeJq7RABK4Eq:-KqSsbREt05fawkPVZBL:bc7f5vr) ([raw file](https://github.com/vaadin/framework/blob/1d4dcdb8abd3c034de54c3d757cc2a1340596df2/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L610)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

probably should also set uiHasValidDateString when the server calls setValue() and everything is ok
</blockquote></details>

This piece of code handles only RangeCheck, and all the parsing stuff should be handled already at this point

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java, line 23 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRUuA7hBVDsU5jBVkN:-KqSoD_botUa60exup1p:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/components/datefield/DateTextHandlingTest.java#L23)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Method added in master - rebasing this change would allow using it.
</blockquote></details>

Done.

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateFieldIsValidTest.java, line 49 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRVrNWGzCbEfeKQFqh:-KqSoGtk1T0w_Rmpk3ou:b-m4ocyp) ([raw file](https://github.com/vaadin/framework/blob/94f2e1f4227c59b21011747ab04ed553cfa235a8/uitest/src/test/java/com/vaadin/tests/components/datefield/DateFieldIsValidTest.java#L49)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Was it an intended change that there is no VCE for invalid value? (My impression is that it was the intended behavior in the past anyway.)
</blockquote></details>

Reverted back

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DateTimeFieldIsValidTest.java, line 49 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRVdGjKpI0_GO2Lrq_:-KqSnmDBkSo6-6Zk28mt:b-m4ocyp) ([raw file](https://github.com/vaadin/framework/blob/94f2e1f4227c59b21011747ab04ed553cfa235a8/uitest/src/test/java/com/vaadin/tests/components/datefield/DateTimeFieldIsValidTest.java#L49)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Was it an intended change that there is no VCE for invalid value? (My impression is that it was the intended behavior in the past anyway.)
</blockquote></details>

Reverted back

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: 4 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 235 at r1](https://reviewable.io:443/reviews/vaadin/framework/9740#-Kq20zZl_6F27DAlO1IH:-KqTY_46f3Us2ap1TDfJ:bb2wrs7) ([raw file](https://github.com/vaadin/framework/blob/0517853905c10f5594adf799d9664f96bb863d78/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L235)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

As far as I understand the form can\t be submitted until internal error is cleared by entering correct date by typing or by choosing from the calendar. Keeping old value in the field shows previous date in the calendar popup even if user typed something incorrect, which is UX improvement, imho. @anasmi, what do you think?
</blockquote></details>

Now null is set to the value.

---

*[uitest/src/test/java/com/vaadin/tests/components/datefield/DisabledParentLayoutTest.java, line 68 at r2](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqRVJCqQp7-hMxB05YR:-KqT_nWwS7vJYlJXaXcB:b-8sbaau) ([raw file](https://github.com/vaadin/framework/blob/63b5f11988e249270e9bec8477f5fe88403a4c7f/uitest/src/test/java/com/vaadin/tests/components/datefield/DisabledParentLayoutTest.java#L68)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

will the TODO be addressed?
</blockquote></details>

#9764

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 1 of 9 files at r2, 1 of 6 files at r3, 3 of 4 files at r4.
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 212 at r4](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqWaTNYvgZA2rrqb6Y9:-KqWaTNYvgZA2rrqb6YA:b-eang68) ([raw file](https://github.com/vaadin/framework/blob/4bf6bd881f663a79a7bad7543366a3c30a9fab43/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L212)):*
> ```Java
>             // this enables analyzing invalid input on the server
>             // this variable is null if the date was chosen with popup calendar
>             // or containds user-typed string
> ```

(containds -> contains - not a blocker)

---

*[server/src/main/java/com/vaadin/ui/AbstractDateField.java, line 222 at r4](https://reviewable.io:443/reviews/vaadin/framework/9740#-KqWbJDVgD6KjO9ju0cj:-KqWbJDVgD6KjO9ju0ck:byj4vm9) ([raw file](https://github.com/vaadin/framework/blob/4bf6bd881f663a79a7bad7543366a3c30a9fab43/server/src/main/java/com/vaadin/ui/AbstractDateField.java#L222)):*
> ```Java
> 
>                 newDate = null;
>                 hasChanges = !uiHasValidDateString;//TODO check if those 3 lines
> ```

(inconsistent whitespace at the beginning of these three comments and could also have just one comment before the three lines to reduce autoformatting changes - not a blocker)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740)*
<!-- Sent from Reviewable.io -->

hesara commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - not perfect but there are tickets for follow-up

---

Reviewed 1 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740#-:-KqX0y-5TAh3qwp8FV7B:b-cp2axu)*
<!-- Sent from Reviewable.io -->

hesara committed:
Updates based on review

tsuoanttila commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9740#-:-KqX2JXpeX_lkIbiDLdU:bnfp4nl)*
<!-- Sent from Reviewable.io -->
